### PR TITLE
[libclc] Add new function implementations in math.

### PR DESCRIPTION
**Pull request by GYDmedwin**
In the math package, there are only atan.cl, cbrt.cl, cos.cl, sin.cl, and sqrt.cl files initially. The other files in the math package I submitted this time are all newly implemented. We are using the SYCL-CTS test set for testing, one of which is the math_builtin_api test. Only when these files are newly added can it be compiled and passed.

### CODE REVIEW

**Comment by GYDmedwin** [comment]
@bader Excuse me to interrupt, can you please see if the new request meets the requirements?

**Comment by GYDmedwin** [comment]
@jchlanda Excuse me to interrupt, can you please see if the new request meets the requirements?

**Comment by GYDmedwin** [comment]
@t4c1 Excuse me to interrupt, can you please see if the new request meets the requirements?

**Review by t4c1 — COMMENTED** [review]
LGTM.

File: libclc/amdgcn-amdhsa/libspirv/math/acos.cl [hunk 1]
```diff
@@ -0,0 +1,19 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <spirv/spirv.h>
+#include <clcmacro.h>
```
Comment (keryell) [inline libclc/amdgcn-amdhsa/libspirv/math/acos.cl:10]: Sorting everywhere?

Comment (GYDmedwin) [reply libclc/amdgcn-amdhsa/libspirv/math/acos.cl:10]: Thanks for your suggestion, I will change it now.

Comment (GYDmedwin) [reply libclc/amdgcn-amdhsa/libspirv/math/acos.cl:9]: @keryell  Sorry to disturb you, I have modified it, please check if it is correct.

Comment (keryell) [reply libclc/amdgcn-amdhsa/libspirv/math/acos.cl:9]: > @keryell Sorry to disturb you, I have modified it, please check if it is correct.

Thanks! It looks quite better to figure out what was added.

**Comment by bader** [comment]
The description doesn't really help reviewers.

> Made some changes based on the last suggestion. 

Could you clarify what suggestion do you mean here?

> Add some new function implementations in the math package, so that the operator can run normally

Which function implementations were added and what is "the operator"? Could you provide more context, please?

**Comment by GYDmedwin** [comment]
@bader Sorry to trouble you, I mean the suggestions made by you and other reviewers in the last request, including removing redundant comment codes, adding open source information in the code, and modifying the newly added names in the SOURCE file Arrange lexicographically, etc.

**Comment by GYDmedwin** [comment]
@bader In the math package, there are only atan.cl, cbrt.cl, cos.cl, sin.cl, and sqrt.cl files initially. The other files in the math package I submitted this time are all newly implemented. We are using the SYCL-CTS test set for testing, one of which is the math_builtin_api test. Only when these files are newly added can it be compiled and passed. I'm sorry for not being able to describe clearly before, thank you for your patience to advise me.

**Comment by bader** [comment]
> the last request

Do you refer to https://github.com/intel/llvm/pull/4849?

**Comment by bader** [comment]
> @bader In the math package, there are only atan.cl, cbrt.cl, cos.cl, sin.cl, and sqrt.cl files initially. The other files in the math package I submitted this time are all newly implemented. We are using the SYCL-CTS test set for testing, one of which is the math_builtin_api test. Only when these files are newly added can it be compiled and passed. I'm sorry for not being able to describe clearly before, thank you for your patience to advise me.

Thanks. Please, add this information to the pull request description.

**Comment by bader** [comment]
@GYDmedwin, could you fix formatting, please?

File: libclc/amdgcn-amdhsa/libspirv/math/ldexp.cl [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#include <spirv/spirv.h>
+
+double __ocml_ldexp_f64(double, int);
+float __ocml_ldexp_f32(float, int);
+
+#define __CLC_FUNCTION __spirv_ocl_ldexp
+#define __CLC_BUILTIN __ocml_ldexp
+#define __CLC_BUILTIN_F __CLC_XCONCAT(__CLC_BUILTIN, _f32)
+#define __CLC_BUILTIN_D __CLC_XCONCAT(__CLC_BUILTIN, _f64)
+
```
Comment (bader) [inline libclc/amdgcn-amdhsa/libspirv/math/ldexp.cl:21]: Are these defines needed?
I suppose `_CLC_DEFINE_BINARY_BUILTIN` does the job.

Please, check if the same defines can be removed from `libclc/amdgcn-amdhsa/libspirv/math/frexp.cl`, `libclc/amdgcn-amdhsa/libspirv/math/modf.cl` and `libclc/amdgcn-amdhsa/libspirv/math/sincos.cl` as well.

File: libclc/amdgcn-amdhsa/libspirv/math/tables.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
+/*
```
Comment (bader) [inline libclc/amdgcn-amdhsa/libspirv/math/tables.h:1]: Could you point out where this header is included? I can find some includes in `generic` directory, but I don't think this header is included there.

I can't find uses for `libclc/amdgcn-amdhsa/libspirv/math/clc_sw_binary.inc` and `libclc/amdgcn-amdhsa/libspirv/math/log_base.h` as well.

File: libclc/amdgcn-amdhsa/libspirv/math/ldexp.cl (hunk 2 above)
Comment (GYDmedwin) [reply libclc/amdgcn-amdhsa/libspirv/math/ldexp.cl:21]: Thank you, I will check now.

File: libclc/amdgcn-amdhsa/libspirv/math/tables.h (hunk 3 above)
Comment (GYDmedwin) [reply libclc/amdgcn-amdhsa/libspirv/math/tables.h:1]: According to your suggestion, I just checked the code. These three files are needed for the new implementation, but there are still some problems, so the new implementation has not been submitted. I will delete them later. They will appear in my next submission. Thank you for your careful inspection.

File: libclc/amdgcn-amdhsa/libspirv/math/sincos.cl [hunk 4]
```diff
@@ -0,0 +1,60 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+  #include <clcmacro.h>
+  #include <spirv/spirv.h>
```
Comment (bader) [inline libclc/amdgcn-amdhsa/libspirv/math/sincos.cl:10]: Interesting... clang-format doesn't complain here - all code except the first comment block is shifted by 2 spaces.

**Review by bader — APPROVED** [review]

**Comment by zjin-lcf** [comment]
I got the following error:

CMake Error at /path/to/llvm/libclc/cmake/modules/AddLibclc.cmake:54 (add_library):
  Cannot find source file:

    /path/to/llvm/libclc/amdgcn-amdhsa/libspirv/math/fma.cl

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .m .M .mm .h .hh .h++ .hm
  .hpp .hxx .in .txx
Call Stack (most recent call first):
  /path/to/llvm/libclc/CMakeLists.txt:368 (add_libclc_builtin_set)

**Comment by GYDmedwin** [comment]
@zjin-lcf Sorry, this is an error, you can delete the file name directly in the SOURCE file.

**Comment by zjin-lcf** [comment]
I appreciate very much your effort of adding many math functions for the HIP support!

**Comment by GYDmedwin** [comment]
@zjin-lcf Thank you for your appreciation, and I am also very grateful to you and other reviewers for your suggestions to me, which allowed me to learn a lot of new knowledge.

**Comment by zjin-lcf** [comment]
@GYDmedwin   I have a question. Are 'fma', 'ilogb', and 'sinh' the last three un-implemented math functions ?

**Comment by GYDmedwin** [comment]
@zjin-lcf In my test，‘ilogb’, ‘log2’, ‘remainder’ the last three un-implemented math functions. I am trying to implement them, but there are still some problems. I did not encounter ‘fma’, it may be my handwriting wrong. The ‘sinh’ source file was dropped, I will upload it now. I am currently repairing ‘ilogb’ and have been testing it. I forgot to delete its name in the SOURCE file. Thank you for your careful inspection, I will correct it immediately.

**Comment by zjin-lcf** [comment]
@GYDmedwin   No problem. I am building the repository with the Python script and observed some errors that abort the compilation.  
Thanks for your great work.

```
In file included from /path/to/llvm/libclc/amdgcn-amdhsa/libspirv/math/ldexp.cl:10:
/path/to/llvm/libclc/generic/include/config.h:23:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_subnormals_disabled();
^
/path/to/llvm/libclc/generic/include/config.h:24:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp16_subnormals_supported();
^
/path/to/llvm/libclc/generic/include/config.h:25:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp32_subnormals_supported();
^
/path/to/llvm/libclc/generic/include/config.h:26:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp64_subnormals_supported();
^
In file included from /path/to/llvm/libclc/amdgcn-amdhsa/libspirv/math/ldexp.cl:11:
In file included from /path/to/llvm/libclc/generic/include/math/math.h:13:
/path/to/llvm/libclc/generic/include/config.h:23:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_subnormals_disabled();
^
/path/to/llvm/libclc/generic/include/config.h:24:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp16_subnormals_supported();
^
/path/to/llvm/libclc/generic/include/config.h:25:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp32_subnormals_supported();
^
/path/to/llvm/libclc/generic/include/config.h:26:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp64_subnormals_supported();
^
In file included from /path/to/llvm/libclc/amdgcn-amdhsa/libspirv/math/ldexp.cl:10:
/path/to/llvm/libclc/generic/include/config.h:23:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_subnormals_disabled();
^
/path/to/llvm/libclc/generic/include/config.h:24:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp16_subnormals_supported();
^
/path/to/llvm/libclc/generic/include/config.h:25:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp32_subnormals_supported();
^
/path/to/llvm/libclc/generic/include/config.h:26:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp64_subnormals_supported();
^
In file included from /path/to/llvm/libclc/amdgcn-amdhsa/libspirv/math/ldexp.cl:11:
In file included from /path/to/llvm/libclc/generic/include/math/math.h:13:
/path/to/llvm/libclc/generic/include/config.h:23:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_subnormals_disabled();
^
/path/to/llvm/libclc/generic/include/config.h:24:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp16_subnormals_supported();
^
/path/to/sycl-space/llvm/libclc/generic/include/config.h:25:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp32_subnormals_supported();
^
/path/to/llvm/libclc/generic/include/config.h:26:1: error: unknown type name '_CLC_DECL'
_CLC_DECL bool __clc_fp64_subnormals_supported();
^
```

**Comment by GYDmedwin** [comment]
@zjin-lcf Sorry, I did not encounter this problem when compiling, I will take time to look at this problem later, thank you.

**Comment by zjin-lcf** [comment]
I will try again. Thanks.

**Comment by zjin-lcf** [comment]
I needed to comment the lines  in the ldexp.cl so that the compile script in Python would not report compile error shown above.
#include <config.h>
#include <math/math.h>

**Comment by zjin-lcf** [comment]
@GYDmedwin 

I tried an example and got the following error. Is it reproducible ?  If not, I will build again.  Thanks.

lld: error: undefined hidden symbol: __spirv_ocl_frexp(float, int*)

**Comment by zjin-lcf** [comment]
The example is https://github.com/zjin-lcf/HeCBench/tree/master/rfs-sycl

**Comment by FMarno** [comment]
@zjin-lcf I've put up a PR here that should fix frexp https://github.com/intel/llvm/pull/5377.
There will likely be similar problems with modf and sincos since they both also use pointers.

**Comment by bader** [comment]
@FMarno, can we fix `modf` and `sincos` in #5377 too?

**Comment by FMarno** [comment]
@bader Sure, I just need to get my head around the macros first!

**Pull request merged**